tls/s2n_recv.c

```diff
@@ -128,6 +129,9 @@ ssize_t s2n_recv(struct s2n_connection *conn, void *buf, ssize_t size, s2n_block
         if (r < 0) {
```

Comment:
I think we'll fail to expire the session cache entry for cases other than S2N_ERR_BLOCKED. We could reach line 147 without calling the cache_delete callback. Could we do something like this: 
```
if (r < 0) {
    if (s2n_errno == S2N_ERR_BLOCKED) {
        ...
    }
    /* Expire entry */
    if (s2n_is_caching_enabled(conn->config) && conn->session_id_len) {
                     conn->config->cache_delete(conn->config->cache_delete_data, conn->session_id, conn->session_id_len);
    }
    /* Handle special closed case */
    if (s2n_errno == S2N_ERR_CLOSED) {
    ...
    }

    /* Some fatal error other than blocked/closed */
    return -1;
}